design-proposals/shadow-node.md

```diff
@@ -0,0 +1,188 @@
+  * Old Virt-API: no upgrade issue.
+* New CRD; new RBAC; new virt-handler; old virt-controller; old virt-API:
+  * new virt-handler: shadowNode is written to but it is not synced to the node yet. However, virt-handler keeps patching the node directly, so there is no upgrade issue.
+  * [same as before] Old controllers: no upgrade issue, virt controller RBAC is not changed.
+  * [same as before] Old Virt-API: no upgrade issue.
+* New CRD; new RBAC; new virt-handler; new virt-controller; old virt-API:
+  * New virt-handler: no upgrade issue. virt-handler patches both node and shadow node.
+  * New controllers: will start getting shadowNode requests once they are created. no upgrade issue.
+  * [same as before] Old Virt-API: no upgrade issue.
+
+### After upgrade rollout:
+Virt-handler will keep trying (and failing) to patch the node since the backup RBACs are now removed. Virt-handler will ignore these cases by ignoring RBAC "forbidden" errors. This behavior should be kept until we no longer support upgrades from the current release.
+
+## Alternative solutions
+### Using status section
+Same proxy solution as primary solution where virt-hadler updates the labels/annotations somewhere, and it gets copied to the node by virt-controller, but instead of using a new CRD as the medium - writing the labels/Annotations to the status section of an existing CRD like kubevirt CR.
+
+Pros:
+- Avoid introducing a new CRD to the project.
+
+Cons:
+- The status section is considered a user facing API section. Dynamically adding/updating 150+ technical labels there may only confuse the user and thus is not advisable.
+
+### Using the virt-handler pod as a medium to pass the labels/annotations
+Same proxy solution as primary solution where virt-handler updates the labels/annotations somewhere, and it gets copied to the node by virt-controller, but instead of using a new CRD as the medium - writing the labels/annotations to the virt-handler pods's labels/annotations.
+Labels/Annotations that need to pass to the node should get a special prefix (like "nodesync/"), so that regular labels don't get passed to the node by mistake.
+
+Pros:
+- Avoid introducing a new CRD to the project.
+- A simple and clean approach.
+
+Cons:
+- It requires giving the virt-handler RBAC permission to update pods, which is generally considered risky.
+- It's only relevant for passing labels/annotations. CRD approach allows for future expansions.
+
+### Creating a GRPC communication between virt-handler and virt-controller
+Pass the labels/annotations through a grpc channel to the virt-controller.
+
+Pros:
+- Avoid introducing a new CRD to the project.
+- security: Virt-handler does not need to access kube-API at all, and cannot modify unrelated nodes or shadow nodes. Virt-controller does all the risky work
+
+Cons:
+- More complex. It's easier and more appealing to use the k8s API, i.e. patching labels/annotations, than to implement and maintain GRPC case logic internally.
```

Comment:
Perhaps I am missing something, but could you please explain what is more complex about it and why a CRD is more appealing?

virt-handler is already serving other gRPC commands for the virt-api, so I see no complexity to add another command to serve virt-controller/virt-operator.

The use of a CRD hints there is a need to expose the information, either for input or output for external projects to use/consume. But the requirement here is internal, nothing to do with an external project consuming this new CRD.

Reply:
plainly, it's more complex and will take more time to implement than the suggested CRD option.

I personally don't see a problem with adding a CRD for internal use (no significant increase in storage, API calls, etc..).
However I agree that adding a new CRD for such a small purpose of proxying info to the node is an overkill, so it justifies looking for an alternative solution. Using GRPC is such a candidate solution.

Reply:
Can you estimate the extra implementation effort in terms of time?

Reply:
well, the effort would comprise of several steps (of course open for suggestion/improvements):
- on virt-handler:
  - replace every place we patch the node with a write to a new "cache", a place where all the new annotations/labels will be stored.
  - create or use an existing grpc/rest server, and use it to return this cache when asked. 
- on virt-controller:
  - add a completely new grpc client component, that will 
    - run only on the leader controller pod (similar to other controllers).
    - run as a go routine and poll the client for updates every Fr minutes. Fr will the lowest of the frequencies, meaning heartbeat (=1m IIRC). 
    - ask every virt-handler server separately (need to make sure it's possible to address them in a unique manner)
  - for each grpc response, thie client will
    - filter only the kubevirt-owned labels/annotations
    - patch them to the appropriate node accordingly.
   
Well, as to your question, I think it will take longer than the proposed solution. How much? maybe 1.5 as long? I would prefer to consult with other virt team members to be more confident.